crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,599 @@
+    unresolved_ptb: UnresolvedProgrammableTransaction,
+) -> Result<ProgrammableTransaction> {
+    let inputs = unresolved_ptb
+        .inputs
+        .into_iter()
+        .enumerate()
+        .map(|(arg_idx, arg)| {
+            resolve_arg(
+                reader,
+                called_packages,
+                &unresolved_ptb.commands,
+                arg,
+                arg_idx,
+            )
+        })
+        .collect::<Result<_>>()?;
+
+    ProgrammableTransaction {
+        inputs,
+        commands: unresolved_ptb
+            .commands
+            .into_iter()
+            .map(TryInto::try_into)
+            .collect::<Result<_, _>>()?,
+    }
+    .pipe(Ok)
+}
+
+fn resolve_arg(
+    reader: &StateReader,
+    called_packages: &HashMap<ObjectId, NormalizedPackage>,
+    commands: &[Command],
+    arg: UnresolvedInputArgument,
+    arg_idx: usize,
+) -> Result<CallArg> {
+    match arg {
+        UnresolvedInputArgument::Pure { value } => CallArg::Pure(value),
+        UnresolvedInputArgument::ImmutableOrOwned(obj_ref) => CallArg::Object(
+            ObjectArg::ImmOrOwnedObject(resolve_object_reference(reader, obj_ref)?),
+        ),
+        UnresolvedInputArgument::Shared {
+            object_id,
+            initial_shared_version: _,
+            mutable: _,
+        } => {
+            let id = object_id.into();
+            let object = reader
+                .inner()
+                .get_object(&id)?
+                .ok_or_else(|| ObjectNotFoundError::new(object_id))?;
+
+            let initial_shared_version = if let sui_types::object::Owner::Shared {
+                initial_shared_version,
+            } = object.owner()
+            {
+                *initial_shared_version
+            } else {
+                return Err(RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    format!("object {object_id} is not a shared object"),
+                ));
+            };
+
+            let mut mutable = false;
+
+            for (command, idx) in find_arg_uses(arg_idx, commands) {
+                match (command, idx) {
+                    (Command::MoveCall(move_call), Some(idx)) => {
+                        let function = called_packages
+                            // Find the package
+                            .get(&move_call.package)
+                            // Find the module
+                            .and_then(|package| {
+                                package.normalized_modules.get(move_call.module.as_str())
+                            })
+                            // Find the function
+                            .and_then(|module| module.functions.get(move_call.function.as_str()))
+                            .ok_or_else(|| {
+                                RestError::new(
+                                    axum::http::StatusCode::BAD_REQUEST,
+                                    "unable to find function",
```

Comment:
would it be useful to specify which function is not found in which command?

Reply:
That's a good point, i'll see what i can do about errors